### PR TITLE
feat: LTI 1.3 Platforms & registration repo implemented

### PR DESCRIPTION
**Pull request by alexeyvy**
## Ticket
https://oat-sa.atlassian.net/browse/TR-879

## Summary
LTI 1.3 Platforms + Registration repo implemented.

## How to test
At the moment, it's not possible to test the whole workflow including the registration repository, as LTI1.3 isn't implemented, yet.
The only way is to check the UI screen is correct. Here's how to do that:
1. Log in to the BO.
2. Click the 'Gear' icon to open preferences, select 'LTI 1.3 Platform' which should show up with this PR applied.
3. Play around with CRUD on platforms etc, check the validation is implemented correctly.

## Sandbox environment
Deployed on http://cgen-ltiregistrations.playground.kitchen.it.taocloud.org:42141/

### CODE REVIEW

**Review by emgolubev — APPROVED** [review]
only code review. Looks good.

**Comment by emgolubev** [comment]
**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Comment by emgolubev** [comment]
Are you sure the target should be `master`?

**Comment by quintanilhar** [comment]
[Last build](http://qa.kitchen.it.taocloud.org/blue/organizations/jenkins/ci-pipeline-cg-extension-pr%2Fextension-tao-lti/detail/PR-297/3/pipeline) has failed, should we be worried about that?

File: models/classes/Platform/Repository/LtiPlatformFactory.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoLti\models\classes\Platform;
```
Comment (wazelin) [inline models/classes/Platform/Repository/LtiPlatformFactory.php:23]: Something is still off ;)

```suggestion
namespace oat\taoLti\models\classes\Platform\Repository;
```

Don't forget to fix the imports where needed afterwards.

File: models/classes/Platform/LtiPlatform.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    public function __construct(
+        string $id,
+        string $label,
+        string $clientId,
+        string $deploymentId,
+        string $audience,
+        string $ouath2AccessTokenUrl,
+        string $oidcAuthenticationUrl,
+        string $jwksUrl
```
Comment (quintanilhar) [inline models/classes/Platform/LtiPlatform.php:62]: I'd only point out the possible inconsistency we can have in our domain layer due to the fact we're relying on primitives only.

What do you think of creating a VO (Url) to ensure we'll always get a well-formed URL from the clients?

**Review by quintanilhar — DISMISSED** [review]
- [X] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Pull request's target is not `master`
- [X] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [X] Commits messages are meaningful
- [X] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Comment by quintanilhar** [comment]
Do we have any other task/user story to document this new capability in our system?

File: models/classes/Platform/Repository/LtiPlatformRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace oat\taoLti\models\classes\Platform\Repository;
+
+use Countable;
+use oat\taoLti\models\classes\Platform\LtiPlatform;
+
+/**
+ * Service methods to manage the LTI platform business objects.
+ */
+interface LtiPlatformRepositoryInterface extends Countable
```
Comment (wazelin) [inline models/classes/Platform/Repository/LtiPlatformRepositoryInterface.php:29]: I'm not quite sure what's the purpose of this interface?
We should probably literally implement https://github.com/oat-sa/lib-lti1p3-core/blob/master/src/Registration/RegistrationRepositoryInterface.php, without introducing an interface of our own, unless we need an adapter implementation for some reason.

**Review by wazelin — CHANGES_REQUESTED** [review]

**Review by quintanilhar — COMMENTED** [review]
👇🏻

File: models/classes/Platform/Repository/LtiPlatformRepositoryInterface.php (hunk 3 above)
Comment (alexeyvy) [reply models/classes/Platform/Repository/LtiPlatformRepositoryInterface.php:29]: It's an interface for Platforms, not Registrations. This interface may be extended in the future with non-RDF implementation with no impact on the Registration repo.

**Comment by alexeyvy** [comment]
>Last build has failed, should we be worried about that?

It fails on `develop` too:

```
vendor/bin/phpunit taoLti/test/unit/models/classes/Security/AuthorizationServerFactoryTest.php
There was 1 error:

1) unit\models\classes\Security\AuthorizationServerFactoryTest::testCreate
LogicException: Unable to read key from file -----BEGIN RSA PRIVATE KEY-----
ABC-----END RSA PRIVATE KEY-----

/var/www/html/vendor/league/oauth2-server/src/CryptKey.php:67
/var/www/html/taoLti/models/classes/Security/AuthorizationServer/AuthorizationServerFactory.php:58
/var/www/html/taoLti/test/unit/models/classes/Security/AuthorizationServerFactoryTest.php:87

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.

```

That bug isn't related to this task, although I tried to look into it quickly, it wasn't obvious to me how to fix it.

@quintanilhar

File: models/classes/Platform/Repository/LtiPlatformFactory.php (hunk 1 above)
Comment (alexeyvy) [reply models/classes/Platform/Repository/LtiPlatformFactory.php]: Fixed, thanks

**Comment by quintanilhar** [comment]
> > Last build has failed, should we be worried about that?
> 
> It fails on `develop` too:
> 
> ```
> vendor/bin/phpunit taoLti/test/unit/models/classes/Security/AuthorizationServerFactoryTest.php
> There was 1 error:
> 
> 1) unit\models\classes\Security\AuthorizationServerFactoryTest::testCreate
> LogicException: Unable to read key from file -----BEGIN RSA PRIVATE KEY-----
> ABC-----END RSA PRIVATE KEY-----
> 
> /var/www/html/vendor/league/oauth2-server/src/CryptKey.php:67
> /var/www/html/taoLti/models/classes/Security/AuthorizationServer/AuthorizationServerFactory.php:58
> /var/www/html/taoLti/test/unit/models/classes/Security/AuthorizationServerFactoryTest.php:87
> 
> ERRORS!
> Tests: 1, Assertions: 0, Errors: 1.
> ```
> 
> That bug isn't related to this task, although I tried to look into it quickly, it wasn't obvious to me how to fix it.
> 
> @quintanilhar

I think it's very important to not leave broken tests in our applications, we need to stop and fix them whenever we detect them. 

I have prepared a small patch to fix it, can you check and commit to your branch if you agree, please? 

[fix__Broken_test_in_AuthorizationServerFactoryTest_php.patch.zip](https://github.com/oat-sa/extension-tao-lti/files/6981402/fix__Broken_test_in_AuthorizationServerFactoryTest_php.patch.zip)

**Review by quintanilhar — APPROVED** [review]

**Pull request merged**